framework/src/materials/Material.C

```diff
@@ -182,12 +187,45 @@ Material::getOutputs()
 }
 
 void
+Material::subdomainSetup()
+{
+  if (_constant_option == 2)
+  {
+    _current_subdomain_id = _current_elem->subdomain_id();
+
+    unsigned int nqp = _fe_problem.getMaxQps();
+
+    MaterialProperties & props = _material_data->props();
+    for (const auto & prop_id : _supplied_prop_ids)
+      props[prop_id]->resize(nqp);
+
+    _qp = 0;
+    computeSubdomainProperties();
+
+    for (const auto & prop_id : _supplied_prop_ids)
+    {
+      for (decltype(nqp) qp = 1; qp < nqp; ++qp)
+        props[prop_id]->qpCopy(qp, props[prop_id], 0);
+    }
+  }
+}
+
+void
+Material::computeSubdomainProperties()
+{
+  mooseError("computeSubdomainQpProperties in Material '", name(), "' needs to be implemented");
+}
+
+void
 Material::computeProperties()
 {
+  if (_constant_option == 2)
```

Comment:
Actually, because I want to keep the name `none`, `element`, `subdomain`, which are not good for native enum (not capitalized, too common), I think the current way is much simpler...